openedx_proversity_reports/reports/time_spent_report_per_user.py

```diff
@@ -0,0 +1,249 @@
+            filter_func=block_type_filter,
+            yield_descendants_of_unyielded=True,
+        )
+        self.course_block_structure = course_block_structure
+
+        return course_component_blocks
+
+
+    def get_google_bigquery_course_id(self):
+        """
+        Return the course id as it is used in Google Big Query for the dataset name.
+        Get the deprecated course key string and then replaces the fordward slash characters '/'
+        with underscore characters '_' e.g. course-v1:edX+DemoX+Demo_Course -> edX-DemoX-Demo_Course
+        """
+        deprecated_course_key = self.course_key._to_deprecated_string()  # pylint: disable=protected-access
+
+        return deprecated_course_key.replace('/', '_')
+
+
+    def get_google_bigquery_data(self):
+        """
+        Return the Google BigQuery data.
+
+        Return:
+            google.cloud.bigquery.job.QueryJob.result() instance.
+        """
+        bigquery_client = get_google_bigquery_api_client()
+        query_job = bigquery_client.query(
+            get_google_bigquery_query(
+                course_dataset_name=self.get_google_bigquery_course_id(),
+                date=self.query_date,
+            ),
+            job_config=get_google_bigquery_job_config(),
+        )
+
+        try:
+            query_job.exception()
+        except GoogleAPIError as api_error:
+            for error_item in api_error.errors:
+                logger.error('Google BigQuery API error: %s', error_item.get('message', ''))
+                return []
+
+        if query_job.errors:
+            for error_item in query_job.errors:
+                logger.error('Google BigQuery query error: %s', error_item.get('message', ''))
+                return []
+
+        return query_job.result()
+
+
+    def generate_report_data(self):
+        """
+        Return the time spent per user report data.
+        """
+        if not (self.course_block_structure or self.course_blocks):
+            return {}
+
+        bigquery_data = list(self.get_google_bigquery_data())
+
+        if not bigquery_data:
+            return {}
+
+        course_blocks = list(self.course_blocks)
+        user_data = []
+
+        for user in self.users:
+            user_course_cohort = get_course_cohort(user=user, course_key=self.course_key)
+            user_course_teams = get_course_teams(membership__user=user, course_id=self.course_key)
+            block_data = []
+
+            for course_block in course_blocks:
+                bigquery_item = {}
+
+                for item_data in bigquery_data:
+                    if item_data.get('username', '') == user.username:
+                        bigquery_item = item_data if course_block.block_id in item_data.get('module_id', '') else {}
+
+                    if bigquery_item:
+                        break
+
+                block_data.append({
+                    'average_time_spent': bigquery_item.get('time_umid30', 0) if bigquery_item else 0,
```

Comment:
That is the name of the table column in Google BigQuery. The table contains only two time-related columns so, if you agree, I will keep it as it is and I will add a inline comment, I mean i think it's not neccesary since there are only two selectable columns and the first one it's not useful so, only exists one possible choice. @andrey-canon

Reply:
but `time_umid5 `is another possible choice, isn't it ?